### DISCUSSION TITLE
Questions on quantum circuit simulation and probability approximation

### DISCUSSION THREAD

**Discussion by ChienKaiMa** [jcmgray/quimb#101] [Q&A] [answered]
According to the book "Quantum Computer Systems: Research for Noisy Intermediate-Scale Quantum Computers" (https://ieeexplore.ieee.org/document/9122477), there are `strong simulation` and `weak simulation`.  I want to compare the ability to perform `strong simulation` on different classical simulators.  To match the definition of `strong simulation`, I found the method `to_dense` in QuiMB can yield accurate simulation results of full statevector.  However, since I'm only curious about part of the statevector (or the probability of some arbitrary states), I'm afraid using `to_dense` is not a fair comparison.  When the qubit number is large, the other simulator I'm experimenting with can still calculate the accurate probability of one or more outcomes.  The following questions are my ideas or concerns for my experiments:
1. Is there a quick function in QuiMB such that I can query the state probability of an arbitrary bitstring?
2. The method `sample` in QuiMB can simulate larger circuits and do the work faster.  I wonder if it is possible to obtain the probability amplitude of some states by using `_sample_conditionals` or other variables used during `sample`.  If it is possible, how accurate is the approximation or how to know how accurate it is?

**Comment by jcmgray** [comment, marked as answer]
Is maybe `Circuit.amplitude` what you are looking for? See [here](https://quimb.readthedocs.io/en/latest/tensor-circuit.html#Forming-the-Target-Tensor-Network) & [here](https://quimb.readthedocs.io/en/latest/_autosummary/quimb.tensor.circuit.html?highlight=amplitude#quimb.tensor.circuit.Circuit.amplitude). It's generally cheaper to compute than either the full state vector and also easier than sampling. If you are investigating performance, you might want to consider some of the [things here](https://quimb.readthedocs.io/en/latest/tensor-circuit.html#Performance-Checklist).

> The method sample in QuiMB can simulate larger circuits and do the work faster. I wonder if it is possible to obtain the probability amplitude of some states by using _sample_conditionals or other variables used during sample. If it is possible, how accurate is the approximation or how to know how accurate it is?

Apart from the threshold `simplify_atol`, which is generally close to the dtype precision, there is no approximation in the sampling.
